src/sinks/file.rs

```diff
@@ -0,0 +1,563 @@
+        fn init(path: PathBuf) -> Self {
+            debug!(message = "opening", file = ?path);
+            let mut options = OpenOptions::new();
+            options.create(true).append(true);
+
+            FileSinkState::OpeningFile(options.open(path))
+        }
+    }
+
+    impl FileSink {
+        pub fn new(path: PathBuf) -> Self {
+            Self {
+                path: path.clone(),
+                state: FileSinkState::init(path),
+            }
+        }
+
+        pub fn new_with_encoding(path: &Path, encoding: Option<Encoding>) -> EmbeddedFileSink {
+            let sink = FileSink::new(path.to_path_buf())
+                .sink_map_err(|err| error!("Terminating the sink due to error: {}", err))
+                .with(move |event| Self::encode_event(event, &encoding));
+
+            Box::new(sink)
+        }
+
+        pub fn poll_file(&mut self) -> Poll<&mut FramedWrite<File, BytesCodec>, io::Error> {
+            loop {
+                match self.state {
+                    FileSinkState::Disconnected => return Err(disconnected()),
+
+                    FileSinkState::FileProvided(ref mut sink) => return Ok(Async::Ready(sink)),
+
+                    FileSinkState::OpeningFile(ref mut open_future) => match open_future.poll() {
+                        Ok(Async::Ready(file)) => {
+                            debug!(message = "provided", file = ?file);
+                            self.state = FileSinkState::FileProvided(FramedWrite::new(
+                                file,
+                                BytesCodec::new(),
+                            ));
+                        }
+                        Ok(Async::NotReady) => return Ok(Async::NotReady),
+                        Err(err) => {
+                            self.state = FileSinkState::Disconnected;
+                            return Err(err);
+                        }
+                    },
+                }
+            }
+        }
+
+        fn encode_event(event: Event, encoding: &Option<Encoding>) -> Result<Bytes, ()> {
+            let log = event.into_log();
+
+            let result = match (encoding, log.is_structured()) {
+                (&Some(Encoding::Json), _) | (_, true) => serde_json::to_vec(&log.all_fields())
+                    .map_err(|e| panic!("Error encoding: {}", e)),
+
+                (&Some(Encoding::Text), _) | (_, false) => Ok(log
+                    .get(&event::MESSAGE)
+                    .map(|v| v.as_bytes().to_vec())
+                    .unwrap_or(Vec::new())),
+            };
+
+            result.map(|mut bytes| {
+                bytes.push(b'\n');
+                Bytes::from(bytes)
+            })
+        }
+    }
+
+    impl Sink for FileSink {
+        type SinkItem = Bytes;
+        type SinkError = io::Error;
+
+        fn start_send(
+            &mut self,
+            line: Self::SinkItem,
+        ) -> StartSend<Self::SinkItem, Self::SinkError> {
+            match self.poll_file() {
```

Comment:
this is a good place to use try_ready so you dont need to have the two extra branches in the match statement.

Reply:
would be great, but unfortunately I change type here:
```
Ok(Async::NotReady) => Ok(AsyncSink::NotReady(line)),
```